how-to/workspace-platform-starter/public/common/views/frame/framed/frame.fdc3.js

```diff
@@ -0,0 +1,116 @@
+const listeners = {};
+const contextCache = {};
+
+/**
+ * Handles the message from the child frame.
+ * @param message message from child frame
+ */
+function handleMessage(message) {
+	switch (message.action) {
+		case 'context-received': {
+			const context = message?.data?.context;
+			const targetContextListeners = listeners[context.type];
+			const allContextListeners = listeners['*'];
+			contextCache['*'] = context;
+			if (context?.type !== undefined) {
+				contextCache[context.type] = context;
+			}
+
+			if (targetContextListeners !== undefined && Array.isArray(targetContextListeners)) {
+				for (const targetListener of targetContextListeners) {
+					targetListener.handler(context, message?.data?.metadata);
+				}
+			}
+
+			if (allContextListeners !== undefined && Array.isArray(allContextListeners)) {
+				for (const allListener of allContextListeners) {
+					allListener.handler(context, message?.data?.metadata);
+				}
+			}
+
+			break;
+		}
+		default: {
+			console.warn(`Unknown action passed from frame: ${message.action}`);
+		}
+	}
+}
+
+/**
+ * Takes a post message event and handles it.
+ * @param event postMessage event that is sent from child frame
+ */
+function handleEvent(event) {
+	console.log('Received event', event);
+	console.log('Acceptable domain', parent.origin);
+	const { data, origin } = event;
+	if (origin === parent.origin) {
+		handleMessage(data);
+	}
+}
+
+window.addEventListener('message', handleEvent, false);
```

Comment:
Could we move this to the top, it gets a bit lost in the middle of the file, and add a comment about what its listening for

Reply:
Yep, originally there as the function it referenced was above it but function hoisting will still manage that.